lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/CompilerV1.scala

```diff
@@ -208,22 +208,28 @@ object CompilerV1 {
         case _        => EitherT.fromEither[Coeval](Left("Only union type can be matched"))
       }
       matchingTypes = cases.flatMap(_.types)
-      _ <- EitherT.cond[Coeval](UNION.eq(u, UNION(matchingTypes.toList.map(CASETYPEREF))),
+      val lastEmpty = cases.last.types.isEmpty
+      _ <- EitherT.cond[Coeval](lastEmpty || UNION.eq(u, UNION(matchingTypes.toList.map(CASETYPEREF))),
```

Comment:
not strict enough. if last is empty, matchching types must be subset of possible types, if not empty, it should be exactly the same set.